### ISSUE TITLE
Client: use caching for recurring same uploads

### ISSUE THREAD

**Issue by W0nderMuffin** [mercedes-benz/sechub#1056]
## Situation
We have a weekly scheduled pipeline which does scans for different modules.
Some of the modules didn't change and a scan is performed for these.

## Wanted
If uploads have e.g. the same hash as a previous scan you could save computing time and power by sending the same result back instead of doing the scan again.

**Comment by de-jcup** [comment]
Interesting idea, but in this case the client would have not to create any SecHub job at all.
So at client side steps would be:
1.1. create zip file
1.2. check `~/.sechub/sechub-client-cache/session.json` available
1.3. not available - go on
1.4. after job with uuid= ${jobUUID} has finished and report available, store result  at ~/.sechub/sechub-client-cache/report-${jobUUID}.json and update json with last scan cache information:
```json
{
        "secHubClientCache" : [
                 {
                          "job" : "${jobUUID}",
                          "sourceZipSha256" : "${sha256-of-source.zip}",                          
                          "report" :"$storedLocation"
                          "reportSha256" : "${sha256-of-source.zip}",
                          "time" :"${cacheCreationTime}"
                 }
        ] 
}
```
1.5 exit 0

Next time for next scan:
2.1.  craete zip file
2.2. check `~/.sechub/sechub-client-cache/session.json` available
2.3. yes - check sha256 inside `secHubClientCache" data ? 
2.4. yes - check cache outdated? 
2.4.1 yes - time out. Remove cache entry, store data and proceed at 1.3
2.4.2 no - not timed out, next step:2.5
2.5 check report sha256 okay? 
2.5.1 yes, next step:2.6
2.5.2  not okay, so report corrupt, Remove cache entry, store data and proceed at 1.3
2.6 reuse existing data (job UUID, report stored at disk)

Caveats: In the described steps before, changes on false positives would NOT TAKE effect

I would suggest, the client should 
- normally do NOT use caching, it should be an additional option e.g. "-use-caching" + "-cache-timeout=..."
- prints a warning when cache is reused (that newer false positives are not handled)

**Comment by sven-dmlr** [comment]
Interesting idea. (and sorry for answering so late)

I would _not_ add such functionality to the client because it adds complexity like:
- home directory is differs on the supported platforms
- as you mentioned: changes to false-positives list are a problem

I could imagine that the SecHub server compares the checksum of the upload with the last upload and then can return the last report and skip the scan itself.
On the other hand profiles and executor settings could have been changed since last scan. Then it would be bad, not to scan.

In summary I would say, that a real scan should be done if one is requested.

Maybe this logic can be added to the CI build. (e.g.: compare git hash with last run - and skip SecHub scan if it did not change)

**Comment by sven-dmlr** [comment]
Closing this issue: Won't implement this at the moment.
